### PR TITLE
Separate parsing into its own Phase

### PR DESCRIPTION
**Pull request by adampauls**
First-time contributor. This is a resurrection of lampepfl/dotty#5613. I've left a few inline comments about any confusion I had along the way. 
Motivations:
* Reuse phase infrastructure (e.g. `-Ystop-after:parser`)
* Make it easier to replace the parser (useful in the REPL for example)

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Phases.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:67]: This was redundant because containsPhase already checks for "all"

File: compiler/src/dotty/tools/dotc/core/Phases.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:110]: Check for isCheckable.

File: compiler/src/dotty/tools/dotc/core/Phases.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:410]: Moved up from FrontEnd

File: compiler/src/dotty/tools/dotc/core/SymDenotations.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/core/SymDenotations.scala:576]: I believe this is the intent of this line.

File: compiler/src/dotty/tools/dotc/transform/Pickler.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/transform/Pickler.scala:118]: This was needed to fix the pickling tests. I don't understand why.

File: docs/docs/contributing/testing.md
Comment (adampauls) [inline docs/docs/contributing/testing.md:168]: Wrote up some advice I got from gitter.

**Comment by smarter** [comment]
> This is a resurrection #5613. 

Cool, /cc @allanrenucci

**Comment by adampauls** [comment]
I rebased so there are only two commits. I fixed a test failure. There's a compilation error that seems to be present on master 
```[error] -- [E007] Type Mismatch Error: /Users/adpauls/sm/git/dotty/scaladoc/src/dotty/tools/scaladoc/tasty/ClassLikeSupport.scala:258:97
[error] 258 |            ParametersList(paramList.params.map(mkParameter(_, memberInfo = memberInfo.paramLists(index))), paramListModifier(paramList.params))```

**Comment by griggt** [comment]
```
[error] -- [E008] Not Found Error: /__w/dotty/dotty/community-build/community-projects/akka/plugins/serialversion-remover-plugin/src/main/scala/akka/Plugin.scala:14:30 
[error] 14 |import dotty.tools.dotc.typer.FrontEnd
[error]    |                              ^^^^^^^^
[error]    |                value FrontEnd is not a member of dotty.tools.dotc.typer
```

There is an open ticket upstream to remove this compiler plugin: https://github.com/akka/akka/issues/30245

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (adampauls) [inline compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:115]: This is a little sketchy, but should silence the error in the community build.

**Comment by adampauls** [comment]
Also, /cc @ biboudis

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:83]: 🤔 probably not a big deal, but would be nice to lift this into some REPL or REPL/test code.

I assume it has something to do with how the REPL splits parsing and typing.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:39]: This seems wrong, no?  Then again no checkfile has failed so perhaps I'm wrong.

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala:22]: Seems needless.

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala:52]: 👨🏼‍🎨 
```suggestion
        report.inform(s"parsing ${unit.source}")
```

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala:17]: "parser"

and I presume it would be best to be:
```suggestion
  override def phaseName: String = Parser.name
```

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala:42]:
```suggestion
    val printer = if (ctx.settings.Xprint.value.contains(Parser.name)) default else typr
```

File: compiler/src/dotty/tools/dotc/typer/Checking.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/Checking.scala:798]: Seems like something that should be a pre-condition even before getting into here - but that's cleanup that can follow-up.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (dwijnand) [inline compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:78]:
```suggestion
        report.inform(s"compiling ${unit.source}")
        val newCtx = ctx.fresh.setPhase(this.start).setCompilationUnit(unit)
```

Or perhaps "typing ${unit.source}"?

**Review by dwijnand — COMMENTED** [review]
I'm no expert, but the diff seems reasonable to me.  Left some non-major comments, questions and suggestions.

File: compiler/src/dotty/tools/dotc/typer/Checking.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/typer/Checking.scala]: This is not necessary, it must have papered over another bug somewhere else early on in my testing. Thanks for pointing it out.

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala]: Thanks, good catch.

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala]: Good point, done.

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala]: Done

File: compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/parsing/ParserPhase.scala]: Removed.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:39]: I lifted that from the original PR. Without it, at least `dotty.tools.dotc.CompilationTests` fail.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:84]: Happy to take suggestions on how to fix this, but it's behind my limited understanding of how the REPL and the compiler interact.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:78]: Done

**Review by dwijnand — APPROVED** [review]
LGTM.

I'm not sure if Nico's just putting names to PRs or if he wants Guilliaume to sign off on merging this.

**Comment by adampauls** [comment]
Thanks for the quick review!

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:39]: Yes, even in case of errors in parsing, we run the typer phase, this is particularly useful for IDEs since it means the presentation compiler is usable even when your code doesn't typecheck.

**Comment by adampauls** [comment]
I'll hold off until merging until @smarter has approved.

**Comment by adampauls** [comment]
Should I continue to wait or just go ahead and merge?

**Comment by smarter** [comment]
I should be able to review this tomorrow.

File: docs/docs/internals/overall-structure.md
Comment (smarter) [inline docs/docs/internals/overall-structure.md:100]:
```suggestion
    List(new Parser) ::             // scanner, parser
```

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (smarter) [inline compiler/src/dotty/tools/dotc/typer/TyperPhase.scala:81]: Rather than using this to indirectly check if we're in the REPL, I suggest defining `class TyperPhase(val addRootImports: Boolean = true)` and setting the parameter to false in ReplCompiler.

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/typer/TyperPhase.scala
Comment (adampauls) [reply compiler/src/dotty/tools/dotc/typer/TyperPhase.scala]: Done

File: docs/docs/internals/overall-structure.md
Comment (adampauls) [reply docs/docs/internals/overall-structure.md]: Done, thanks

**Review by smarter — APPROVED** [review]

**Comment by som-snytt** [comment]
woo-hoo

**Comment by fommil** [comment]
@smarter does this mean we can have pre-typer plugins now? 😬  As per https://contributors.scala-lang.org/t/pre-typer-syntactic-plugins-in-scala-3/5565

**Comment by smarter** [comment]
No, pre-typer plugin are still intentionally restricted to research plugins.

**Pull request merged**
